### PR TITLE
sonix sn32f2xx platform support

### PR DESCRIPTION
**Pull request by dexter93**
Includes https://github.com/qmk/qmk_firmware/pull/16339

Requires https://github.com/ChibiOS/ChibiOS-Contrib/pull/317

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: platforms/chibios/bootloaders/sn32_dfu.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#       define BOOTLOADER_MAGIC 0xDEADBEEF
+#       define MAGIC_ADDR (unsigned long *)(SYMVAL(__ram0_end__) - 4)
+
+__attribute__((weak)) void bootloader_jump(void) {
+    *MAGIC_ADDR = BOOTLOADER_MAGIC;  // set magic flag => reset handler will jump into boot loader
+    // Wait for memory to be set before the reset
+    // FIXME: update to wait funcion
+    for(volatile uint32_t i = 0; i < 32; i++){
+        __NOP();
+    }
```
Comment (KarlK90) [inline platforms/chibios/bootloaders/sn32_dfu.c:33]: Should use the `wait_us()` / `wait_ms()` function right?

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Great work! :muscle:

Documentation changes for supported MCUs are missing and AFAIK every new supported platform on QMK requires a corresponding onekey keyboard in the handwired folder.

File: platforms/chibios/bootloaders/sn32_dfu.c (hunk 1 above)
Comment (dexter93) [reply platforms/chibios/bootloaders/sn32_dfu.c:33]: probably, we just now got the wait_ functions working reliably. Still not sure how to translate that to us value

**Comment by dexter93** [comment]
Thanks! Docs will be added. Downside on that is that most of the keyboards on the platform require non-standard matrix setup, will follow that with a new commit soon.
> Great work! muscle
> 
> Nits so far: Documentation changes for supported MCUs are missing and AFAIK every new supported platform on QMK requires a corresponding onekey keyboard in the handwired folder.

File: platforms/chibios/bootloaders/sn32_dfu.c (hunk 1 above)
Comment (dexter93) [reply platforms/chibios/bootloaders/sn32_dfu.c]: 1us seems good enough. updated

**Comment by tzarc** [comment]
Have created a PR -- https://github.com/dexter93/qmk_firmware/pull/10 -- which does some cleanup and removes the unused emulated EEPROM stuff.

Still needs some handwired/onekey variants created.

**Comment by tzarc** [comment]
Original author has gone on sabbatical or equivalent. Suggest someone else raises a PR once it's ready.

**Pull request closed without merging**